ipaserver/install/ipa_cert_fix.py

```diff
@@ -0,0 +1,267 @@
+    if cert.not_valid_after <= now:
+        certs.append((IPACertType.IPARA, cert))
+
+    # Apache HTTPD
+    db = NSSDatabase(nssdir=paths.HTTPD_ALIAS_DIR)
+    cert = db.get_cert('Server-Cert')
+    if cert.not_valid_after <= now:
+        certs.append((IPACertType.HTTPS, cert))
+
+    # LDAPS
+    ds_dbdir = dsinstance.config_dirname(
+        installutils.realm_to_serverid(api.env.realm))
+    db = NSSDatabase(nssdir=ds_dbdir)
+    cert = db.get_cert('Server-Cert')
+    if cert.not_valid_after <= now:
+        certs.append((IPACertType.LDAPS, cert))
+
+    # KDC
+    cert = x509.load_certificate_from_file(paths.KDC_CERT)
+    if cert.not_valid_after <= now:
+        certs.append((IPACertType.KDC, cert))
+
+    return certs
+
+
+def print_intentions(dogtag_certs, ipa_certs):
+    print("The following certificates will be renewed: ")
+    print()
+
+    for certid, cert in dogtag_certs:
+        print_cert_info("Dogtag", certid, cert)
+
+    for certtype, cert in ipa_certs:
+        print_cert_info("IPA", certtype.value, cert)
+
+
+def print_cert_info(context, desc, cert):
+    print("{} {} certificate:".format(context, desc))
+    print("  Subject: {}".format(DN(cert.subject)))
+    print("  Serial:  {}".format(cert.serial_number))
+    print("  Expires: {}".format(cert.not_valid_after))
+    print()
+
+
+def run_cert_fix(certs, extra_certs):
+    ldapi_path = (
+        paths.SLAPD_INSTANCE_SOCKET_TEMPLATE
+        % '-'.join(api.env.realm.split('.'))
+    )
+    cmd = [
+        'pki-server',
+        'cert-fix',
+        '--ldapi-socket', ldapi_path,
+        '--agent-uid', 'ipara',
+    ]
+    for certid, _cert in certs:
+        cmd.extend(['--cert', certid])
+    for _certtype, cert in extra_certs:
+        cmd.extend(['--extra-cert', str(cert.serial_number)])
+    ipautil.run(cmd, raiseonerr=True)
+
+
+def replicate_dogtag_certs(subject_base, ca_subject_dn, certs):
+    for certid, _oldcert in certs:
+        cert_path = "/etc/pki/pki-tomcat/certs/{}.crt".format(certid)
+        cert = x509.load_certificate_from_file(cert_path)
+        print_cert_info("Renewed Dogtag", certid, cert)
+        replicate_cert(subject_base, ca_subject_dn, cert)
+
+
+def install_ipa_certs(subject_base, ca_subject_dn, certs):
+    """Print details and install renewed IPA certificates."""
+    for certtype, oldcert in certs:
+        cert_path = "/etc/pki/pki-tomcat/certs/{}-renewed.crt" \
+                .format(oldcert.serial_number)
+        cert = x509.load_certificate_from_file(cert_path)
+        print_cert_info("Renewed IPA", certtype.value, cert)
+
+        if certtype is IPACertType.IPARA:
+            shutil.copyfile(cert_path, paths.RA_AGENT_PEM)
+            cainstance.update_people_entry(cert)
+            replicate_cert(subject_base, ca_subject_dn, cert)
+        elif certtype is IPACertType.HTTPS:
+            db = NSSDatabase(nssdir=paths.HTTPD_ALIAS_DIR)
+            db.import_pem_cert('Server-Cert', EMPTY_TRUST_FLAGS, cert_path)
```

Comment:
At this point there are 2 Server-Cert certificates for httpd. The previous one should be removed from the NSS DB.

Reply:
Updated the PR to remove the previous cert (same for DS cert below).